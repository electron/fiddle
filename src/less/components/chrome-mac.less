.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 24px;
    font-size: 1.4rem;
    line-height: 15px;

    small,
    p,
    span {
      user-select: none;
    }
  }

  .big_sur_or_later {
    height: 30px !important;
  }
}
