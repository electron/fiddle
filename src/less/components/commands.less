@import "../variables.less";

header {
  box-shadow: 0 0 0 1px rgba(16, 22, 26, 0.2), 0 0 0 rgba(16, 22, 26, 0), 0 1px 1px rgba(16, 22, 26, 0.4);
  background-color: @background-3;
  -webkit-app-region: drag;
}

.disabled-menu-tooltip {
  display: flex;
  width: 100%;

  .bp3-popover-target {
    width: 100%;
  }
}

.commands.is-mac {
  padding:
    10px
    calc(100vw - (env(titlebar-area-x, 0) + env(titlebar-area-width, 0)))
    10px
    env(titlebar-area-x, 0);
}

.commands {
  display: flex;
  justify-content: space-between;
  height: 50px;
  padding:
    10px;
  box-sizing: border-box;
  font-family: @fonts-common;

  button,
  input,
  select {
    -webkit-app-region: no-drag;
    margin-bottom: 0;
    white-space: nowrap;
  }

  select {
    width: 100px;
    margin-left: 5px;
  }

  fieldset {
    border: 0;
    padding: 0;
    margin: 0;
    min-width: 0;
  }

  & > div {
    display: flex;

    &:last-of-type {
      margin-right: 10px;

      .bp3-button-group {
        margin-left: 5px;
      }
    }
  }

  .bp3-control-group {
    margin-right: 0.5rem;
  }

  .address-bar {
    width: 420px;
    transition: width 0.4s;

    input,
    span {
      -webkit-app-region: no-drag;
    }

    &.empty {
      width: 300px;
    }

    &:focus-within {
      span,
      input {
        border-color: @foreground-2;
      }
    }
  }

  & > .title {
    align-self: stretch;
    flex-shrink: 0;

    flex-flow: column nowrap;
    justify-content: center;
    user-select: none;

    margin: 0 10px;
  }
}
