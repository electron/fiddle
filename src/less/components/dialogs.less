@import "../variables.less";

.dialogs {
  z-index: 10;
  position: fixed;
  height: 100vh;
  width: 100vw;
  pointer-events: none;

  * {
    pointer-events: all;
  }
}

.dialog {
  background-color: @background-3;
  border: @border;
  color: @text-color-1;
  border-radius: 6px;
  padding: 10px;
  z-index: 100;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  animation: slide-up 0.4s ease;

  p {
    margin-bottom: 1.25rem;
    font-size: 1.25rem;
  }

  code {
    font-size: 12px;
    background: rgba(255, 255, 255, 0.1);
    padding: 2px;
    line-height: 12px;
  }

  &-buttons {
    margin-top: 1.25rem;
  }

  &.centered {
    position: absolute;
    width: 300px;
    left: calc(50% - (300px / 2));
    top: 100px;
  }

  &.arrow::after,
  &.arrow::before {
    border: solid transparent;
    content: " ";
    height: 0;
    width: 0;
    position: absolute;
    pointer-events: none;
  }

  &.arrow::after {
    border-color: rgba(136, 183, 213, 0);
    border-width: 9px;
  }

  &.arrow::before {
    border-color: rgba(194, 225, 245, 0);
    border-width: 10px;
  }

  &.arrow-top::after,
  &.arrow-top::before {
    bottom: 100%;
    left: 50%;
  }

  &.arrow-top::after {
    border-bottom-color: @background-3;
    margin-left: -9px;
  }

  &.arrow-top::before {
    border-bottom-color: @border-color-1;
    margin-left: -10px;
  }

  &.arrow-bottom::after,
  &.arrow-bottom::before {
    top: 100%;
    left: 50%;
  }

  &.arrow-bottom::after {
    border-top-color: @background-3;
    margin-left: -9px;
  }

  &.arrow-bottom::before {
    border-top-color: @border-color-1;
    margin-left: -10px;
  }

  &.arrow-left::after,
  &.arrow-left::before {
    right: 100%;
    top: 20px;
  }

  &.arrow-left::after {
    border-right-color: @background-3;
    margin-top: -9px;
  }

  &.arrow-left::before {
    border-right-color: @border-color-1;
    margin-top: -10px;
  }

  &.arrow-right::after,
  &.arrow-right::before {
    left: 100%;
    top: 20px;
  }

  &.arrow-right::after {
    border-left-color: @background-3;
    margin-top: -9px;
  }

  &.arrow-right::before {
    border-left-color: @border-color-1;
    margin-top: -10px;
  }

  button {
    color: #1e2527;
    float: right;
    margin-left: 9px;
    margin-bottom: 0;
  }

  a {
    cursor: pointer;
  }

  h4 {
    margin-bottom: 0;
  }

  p:last-of-type {
    margin-bottom: 0;
  }

  input {
    color: rgba(159, 234, 249, 0.5);
    height: 28px;
    margin-top: 10px;
    background-color: @background-1;
    border: 1px solid @border-color-2;
    outline: 0 solid transparent;

    &:focus {
      border-color: @foreground-2;
    }
  }
}

.dialogDrop {
  position: fixed;
  z-index: 100;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  background: black;
  transition: opacity 0.4s ease-in-out;
  pointer-events: all;
  opacity: 0.5;
}

.dialog-add-version {
  .bp3-file-input {
    width: 100%;
  }

  .bp3-callout {
    margin-top: 1rem;
  }
}

select {
  font-size: 12px;
  height: 28px;
  background-color: white;
  line-height: 15px;
}

@keyframes slide-up {
  0% {
    opacity: 0;
    transform: translateY(20px);
  }

  100% {
    opacity: 1;
    transform: translateY(0);
  }
}
