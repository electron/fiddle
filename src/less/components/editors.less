.mosaic-tile {
  animation: fadein 0.15s;
}

.mosaic-window-toolbar  {
  & > div {
    display: flex;
    justify-content: space-between;
    width: 100%;
    padding: 0 5px 0 5px;

    .mosaic-controls {
      transform: scale(0.75);
      transform-origin: right;
    }

    h5 {
      margin: 4px 0 0 0;
    }

    button {
      margin-left: 8px;
    }
  }
}

// TODO: support new file
// update if list of Editor ID changes
@editor-ids: main\.js, renderer\.js, index\.html, preload\.js, styles\.css;

// takes each editor ID and increase its z-index if the parent Mosaic root
// has focused__id class for that specific id.
each(@editor-ids, {
  .focused__@{value}.mosaic .mosaic-window.@{value} {
    z-index: 2;
  }
});

.mosaic-window {
  z-index: 1;
}

@keyframes fadein {
  from { opacity: 0; }
  to { opacity: 1; }
}
