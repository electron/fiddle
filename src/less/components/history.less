.revision-list {
  max-height: 400px;
  overflow-y: auto;
}

.revision-list ul {
  list-style-type: none;
  padding: 0;
  margin: 0;
}

.revision-item {
  padding: 10px;
  margin-bottom: 5px;
  border-radius: 3px;
  cursor: pointer;
  transition: background-color 0.2s ease;
  border-left: 3px solid #106ba3;
}

.revision-item:hover {
  background-color: rgba(167, 182, 194, 0.3);
}

.revision-content {
  display: flex;
  flex-direction: column;
}

.revision-icon {
  margin-right: 8px;
}

.sha-label {
  font-size: 12px;
  color: #738694;
  margin-left: 10px;
  font-family: monospace;
}

.revision-details {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-top: 5px;
}

.revision-date {
  color: #738694;
  font-size: 12px;
}

.revision-changes {
  display: flex;
  gap: 5px;
}

.history-loading {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px;
}
