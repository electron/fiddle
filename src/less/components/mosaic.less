@import '../variables.less';

.fiddle {
  .mosaic-window-toolbar {
    background: var(--background-1);
    box-shadow: inset 0 0 100px 100px rgba(53, 53, 53, 0.1);
    border-top: 1px solid rgba(255, 255, 255, 0.1);
  }

  .mosaic-window-toolbar.draggable:hover {
    box-shadow: inset 0 0 100px 100px rgba(255, 255, 255, 0.1);
  }

  .mosaic-window-title {
    color: var(--text-color-3);
  }

  button.mosaic-default-control {
    padding: 0;
    height: 20px;
    width: 20px;
    margin: 5px;
    background-color: #314246;
    border: 0.1rem solid #222b35;
  }

  button.mosaic-default-control:hover {
    background-color: #131d1f;
  }

  .mosaic-window-body {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    align-items: stretch;
    align-content: stretch;
    overflow: visible;
    z-index: 4;
    background-color: @background-4;
  }

  .mosaic.mosaic-blueprint-theme {
    background: #bdbdbd15;
  }

  .editorContainer,
  .monaco-editor {
    width: 100%;
    height: 100%;
  }

  .editorContainer {
    background: @background-1;
  }
}
