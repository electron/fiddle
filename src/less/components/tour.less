.tour {
  position: fixed;
  top: 0;
  left: 0;
  height: 100vh;
  width: 100vw;
  z-index: 200;
}

.tour-portal {
  z-index: 1000;
}
