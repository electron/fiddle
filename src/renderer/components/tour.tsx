import * as React from 'react';

import { Button, Classes, Dialog } from '@blueprintjs/core';

import { positionForRect } from '../../utils/position-for-rect';

export interface TourScriptStep {
  name: string;
  selector: string;
  title: string;
  content: JSX.Element;
  getButtons?: (handlers: TourStepGetButtonParams) => Array<JSX.Element>;
}

interface TourProps {
  tour: Set<TourScriptStep>;
  onStop: () => void;
}

interface TourState {
  tour: IterableIterator<Array<TourScriptStep>>;
  step: TourScriptStep | null;
  i: number;
}

export interface TourStepGetButtonParams {
  stop: () => void;
  advance: () => void;
}

export class Tour extends React.Component<TourProps, TourState> {
  private resizeHandle = 0;

  constructor(props: TourProps) {
    super(props);

    this.advance = this.advance.bind(this);
    this.stop = this.stop.bind(this);
    this.onResize = this.onResize.bind(this);

    this.state = {
      tour: props.tour.entries(),
      step: null,
      i: 0,
    };
  }

  /**
   * Handles a resize of the window.
   */
  public onResize() {
    if (!this.resizeHandle) {
      this.resizeHandle = window.setTimeout(() => {
        this.forceUpdate();
        this.resizeHandle = 0;
      }, 100);
    }
  }

  /**
   * Component was mounted: Sign up for the window's "resize"
   * event and move to the first step of the tour.
   */
  public componentDidMount() {
    this.advance();

    window.addEventListener('resize', this.onResize);
  }

  public componentWillUnmount() {
    window.removeEventListener('resize', this.onResize);
  }

  /**
   * Moves the tour to the next step
   */
  public advance() {
    const { done, value } = this.state.tour.next();
    const step = done ? null : value[0];

    this.setState({ step, i: this.state.i + 1 });
  }

  /**
   * Stops the tour
   */
  public stop() {
    this.props.onStop();
  }

  public render() {
    const { step } = this.state;

    if (!step) return null;

    return <div className="tour">{this.getStep(step)}</div>;
  }

  /**
   * Return buttons for the dialog for the current step
   * of the tour. By default, we just return a continue/stop
   * combo
   *
   * @param {TourScriptStep} { getButtons }
   * @returns {Array<JSX.Element>}
   */
  private getButtons({ getButtons }: TourScriptStep): Array<JSX.Element> {
    // Did the step bring its own buttons?
    if (getButtons) {
      return getButtons({
        stop: this.stop,
        advance: this.advance,
      });
    }

    // No? Fine! Are we at the end of the tour?
    return this.props.tour.size === this.state.i
      ? [
          <Button
            icon="tick-circle"
            onClick={this.stop}
            key="btn-stop"
            text="Finish Tour"
          />,
        ]
      : [
          <Button
            icon="stop"
            key="btn-stop"
            onClick={this.stop}
            text="Stop Tour"
          />,
          <Button
            icon="step-forward"
            key="btn-adv"
            onClick={this.advance}
            text="Continue"
          />,
        ];
  }

  /**
   * Renders the dialog for the current step of the tour.
   *
   * @param {TourScriptStep} step
   * @param {ClientRect} rect
   * @returns {JSX.Element}
   */
  private getDialogForStep(
    step: TourScriptStep,
    rect: ClientRect,
  ): JSX.Element {
    const buttons = this.getButtons(step);
    const size = { width: 400, height: 300 };
    const margin = 10;
    const position = positionForRect(rect, size);
    const style: React.CSSProperties = {
      position: 'absolute',
      top: position.top,
      left: position.left,
      width: size.width - margin,
      margin: `${margin}px`,
      zIndex: 1000,
    };

    return (
      <Dialog
        key={step.name}
        isOpen={true}
        style={style}
        portalClassName="tour-portal"
        backdropProps={{ style: { visibility: 'hidden' } }}
      >
        <div className={Classes.DIALOG_HEADER}>
          <h4 className={Classes.HEADING}>{step.title}</h4>
        </div>
        <div className={Classes.DIALOG_BODY}>{step.content}</div>
        <div className={Classes.DIALOG_FOOTER}>
          <div className={Classes.DIALOG_FOOTER_ACTIONS}>{buttons}</div>
        </div>
      </Dialog>
    );
  }

  /**
   * Returns the "mask" for a given step.
   *
   * @param {TourScriptStep} step
   * @returns {(JSX.Element | null)}
   */
  private getStep(step: TourScriptStep): JSX.Element | null {
    const { selector } = step;
    const element = document.querySelector(selector) || document.body!;
    const rect = element.getBoundingClientRect();
    const { top, left, width, height } = rect;

    return (
      <>
        {this.getDialogForStep(step, rect)}
        <svg height="100%" width="100%">
          <rect
            fill="rgba(0, 0, 0, 0.65)"
            x="0"
            y="0"
            mask="url(#mask)"
            height="100%"
            width="100%"
          />
          <mask
            id="mask"
            maskUnits="userSpaceOnUse"
            maskContentUnits="userSpaceOnUse"
          >
            <rect
              className="bg"
              x="0"
              y="0"
              fill="white"
              height="100%"
              width="100%"
            />
            <rect
              x={left}
              y={top}
              width={width}
              height={height}
              fill="black"
              rx="5"
              ry="5"
              strokeWidth="5"
              stroke="black"
            />
          </mask>
        </svg>
      </>
    );
  }
}
