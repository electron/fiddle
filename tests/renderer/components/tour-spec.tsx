import * as React from 'react';

import { mount, shallow } from 'enzyme';

import { Tour } from '../../../src/renderer/components/tour';
import { overridePlatform, resetPlatform } from '../../utils';

describe('VersionChooser component', () => {
  const oldQuerySelector = document.querySelector;
  const mockTour = new Set([
    {
      name: 'mock-step-1',
      selector: 'div.mock-1',
      title: 'Step 1',
      content: <span key="1">mock-step-1</span>,
    },
    {
      name: 'mock-step-2',
      selector: 'div.mock-2',
      title: 'Step 2',
      content: <span key="2">mock-step-2</span>,
    },
  ]);

  beforeAll(() => {
    overridePlatform('darwin');
  });

  afterAll(() => {
    resetPlatform();
  });

  beforeEach(() => {
    document.querySelector = jest.fn(() => ({
      getBoundingClientRect: jest.fn(() => ({
        top: 20,
        left: 25,
        height: 120,
        width: 130,
      })),
    }));
  });

  afterEach(() => {
    document.querySelector = oldQuerySelector;
  });

  it('renders', () => {
    const mockStop = jest.fn();
    const wrapper = shallow(<Tour tour={mockTour} onStop={mockStop} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders supplied buttons', () => {
    mockTour.forEach((item) => {
      (item as any).getButtons = () => [<button key="hello">Hello</button>];
    });

    const mockStop = jest.fn();
    const wrapper = shallow(<Tour tour={mockTour} onStop={mockStop} />);

    expect(wrapper.find('button').text()).toBe('Hello');
  });

  it('renders "Finish Tour" at the end', () => {
    const singleItemTour = new Set([
      {
        name: 'mock-step-1',
        selector: 'div.mock-1',
        title: 'Step 1',
        content: <span key="1">mock-step-1</span>,
      },
    ]);

    const mockStop = jest.fn();
    const wrapper = mount(<Tour tour={singleItemTour} onStop={mockStop} />);

    expect(wrapper.find('button').text()).toBe('tick-circleFinish Tour');
  });

  it('handles a missing target', () => {
    (document.querySelector as jest.Mock).mockReturnValueOnce(null);

    const mockStop = jest.fn();
    const wrapper = shallow(<Tour tour={mockTour} onStop={mockStop} />);

    expect(wrapper.find('svg').length).toBe(1);
  });

  it('stops on stop()', () => {
    const mockStop = jest.fn();
    const wrapper = shallow(<Tour tour={mockTour} onStop={mockStop} />);
    const instance: any = wrapper.instance();

    instance.stop();

    expect(mockStop).toHaveBeenCalled();
  });

  it('advances on advance()', () => {
    const mockStop = jest.fn();
    const wrapper = shallow(<Tour tour={mockTour} onStop={mockStop} />);
    const instance: any = wrapper.instance();

    instance.advance();
    expect((wrapper.state('step') as any).name).toBe('mock-step-2');

    instance.advance();
    expect(wrapper.state('step')).toBe(null);
  });

  it('handles a resize', () => {
    jest.useFakeTimers();

    const mockStop = jest.fn();
    const wrapper = shallow(<Tour tour={mockTour} onStop={mockStop} />);
    const instance: any = wrapper.instance();

    instance.forceUpdate = jest.fn();
    instance.onResize();

    expect(instance.resizeHandle).toBeTruthy();
    jest.runAllTimers();

    expect(instance.forceUpdate).toHaveBeenCalled();
  });

  it('removes the resize listener', () => {
    const oldRemove = window.removeEventListener;
    window.removeEventListener = jest.fn();

    const mockStop = jest.fn();
    const wrapper = shallow(<Tour tour={mockTour} onStop={mockStop} />);
    const instance: any = wrapper.instance() as any;

    instance.componentWillUnmount();
    expect(window.removeEventListener).toHaveBeenCalled();

    window.removeEventListener = oldRemove;
  });
});
