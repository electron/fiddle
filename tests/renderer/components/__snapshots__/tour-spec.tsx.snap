// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VersionChooser component renders 1`] = `
<div
  className="tour"
>
  <Blueprint4.Dialog
    backdropProps={
      Object {
        "style": Object {
          "visibility": "hidden",
        },
      }
    }
    canOutsideClickClose={true}
    isOpen={true}
    key="mock-step-1"
    portalClassName="tour-portal"
    style={
      Object {
        "left": 165,
        "margin": "10px",
        "position": "absolute",
        "top": 5,
        "width": 390,
        "zIndex": 1000,
      }
    }
  >
    <div
      className="bp4-dialog-header"
    >
      <h4
        className="bp4-heading"
      >
        Step 1
      </h4>
    </div>
    <div
      className="bp4-dialog-body"
    >
      <span
        key="1"
      >
        mock-step-1
      </span>
    </div>
    <div
      className="bp4-dialog-footer"
    >
      <div
        className="bp4-dialog-footer-actions"
      >
        <Blueprint4.Button
          icon="stop"
          key="btn-stop"
          onClick={[Function]}
          text="Stop Tour"
        />
        <Blueprint4.Button
          icon="step-forward"
          key="btn-adv"
          onClick={[Function]}
          text="Continue"
        />
      </div>
    </div>
  </Blueprint4.Dialog>
  <svg
    height="100%"
    width="100%"
  >
    <rect
      fill="rgba(0, 0, 0, 0.65)"
      height="100%"
      mask="url(#mask)"
      width="100%"
      x="0"
      y="0"
    />
    <mask
      id="mask"
      maskContentUnits="userSpaceOnUse"
      maskUnits="userSpaceOnUse"
    >
      <rect
        className="bg"
        fill="white"
        height="100%"
        width="100%"
        x="0"
        y="0"
      />
      <rect
        fill="black"
        height={120}
        rx="5"
        ry="5"
        stroke="black"
        strokeWidth="5"
        width={130}
        x={25}
        y={20}
      />
    </mask>
  </svg>
</div>
`;
